### ISSUE TITLE
geom_sf() and facet_*(..., scales="free")

### ISSUE THREAD

**Issue by adrfantini** [tidyverse/ggplot2#2651]
Setting `scales="free"` in `facet_*` resets `geom_sf` coordinates handling, even if not necessary. See reprex below.

```r
library(ggplot2)
library(dplyr)
library(sf)
nc <- sf::st_read(system.file("shape/nc.shp", package = "sf"), quiet = TRUE) %>% mutate(dummy=1)
nc2 <- sf::st_read(system.file("shape/nc.shp", package = "sf"), quiet = TRUE) %>% mutate(dummy=2)
#Correct form factor for the plot:
ggplot(rbind(nc, nc2)) +  geom_sf(aes(fill = AREA)) + facet_wrap(~dummy)
#Adding scales="free" causes the plot to be warped:
ggplot(rbind(nc, nc2)) +  geom_sf(aes(fill = AREA)) + facet_wrap(~dummy, scales="free")
```
Without `scales="free"`:
![spectacle twz916](https://user-images.githubusercontent.com/12826141/40350553-6ed0d404-5daa-11e8-92de-e5ca2dd28e10.png)
With `scales="free"`:
![spectacle hcf916](https://user-images.githubusercontent.com/12826141/40350583-859cfc8a-5daa-11e8-8813-c56828e9a894.png)

`scales="free"` should allow facets to have different ranges without altering the plot aspect ratio.

**Comment by adrfantini** [comment]
> It's not generally possible to scale axes freely and maintain the correct aspect ratio and keep facet sizes equal.

What if I want to show some data over USA and Canada, in two different facets? Currently,  (without `scales = "free"`) I'll have two big facets spanning the whole of North America (One with Canada data, one with USA data), instead (using `scales = "free"`) of two smaller facets spanning USA and Canada separately. Why shouldn't this be allowed?

**Comment by clauswilke** [comment]
The problem is that throughout the ggplot2 code base, the assumption is made that axes can be trained independently of other axes. To implement what you're requesting, one would have to write a training algorithm that trains x and y axes simultaneously while producing a targeted aspect ratio. This is possible in principle, but it's not immediately obvious how to do it in the current code base, at least to me. The best way forward might be to write a new faceting function specifically for this purpose, probably first in the form of an extension package. @thomasp85 may have further insight.

**Comment by batpigandme** [comment]
There are ggplot2 extensions <http://www.ggplot2-exts.org/gallery/> that should allow you to arrange two maps side by side as such, e.g. [patchwork](https://github.com/thomasp85/patchwork#patchwork-), and (@clauswilke, correct me if I'm wrong) [cowplots](https://github.com/wilkelab/cowplot).

**Comment by clauswilke** [comment]
@batpigandme Yes, that's the other possible approach, make the plots separately and then arrange side-by-side. But then we don't get the power of faceting.

**Comment by batpigandme** [comment]
Oh yeah, of course, I was just offering a workaround!

**Comment by hadley** [comment]
Yes, this is a bug - this plot should throw an error message.

**Comment by adrfantini** [comment]
Thanks for the bugfix. I do not know the internals of `ggplot` so my request for clarification might seem trivial (and maybe already answered by @clauswilke ), but if the following works, why can't the same be made working with `coord_sf`?

```R
d = data.frame(x=c(1,100), y=c(1,100), f=c(1, 100))
ggplot(d, aes(x=y, y=y)) + geom_point() + facet_wrap(~f, scales="free")
```

**Comment by clauswilke** [comment]
`coord_sf` has a fixed aspect ratio. The appropriate comparison would be something like this:
``` r
library(ggplot2)
d = data.frame(x=c(1,100), y=c(1,100), f=c(1, 100))
ggplot(d, aes(x=y, y=y)) + geom_point() + facet_wrap(~f, scales="free") + coord_fixed()
#> Error: coord_fixed doesn't support free scales
```

![](https://i.imgur.com/p4ajqZq.png)

Created on 2018-05-23 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by adrfantini** [comment]
@clauswilke Thanks for the explanation! I believe this can be closed, right?

**Comment by clauswilke** [comment]
The issue will be automatically closed once the pull request is integrated into the main code base. That hasn’t happened yet.
